lib/cfg.js

```diff
@@ -0,0 +1,368 @@
+            --cbb.end;
+        }
+                
+        return asmArr[cbb.end-1].text;
+    };
+    
+    var generateName = function(name, suffix){
+        var pos = name.indexOf("@");
+        if(pos == -1)
+            return name + "@" + suffix;
+        
+        return name.substring(0, pos+1)+ suffix;
+    };
+
+    // note: x.end-1 possible value: jmp .L*, {jne,je,jg,...} .L*, ret/rep ret, call and any other instruction that doesn't change control flow
+
+    //for(var x of arrOfCanonicalBasicBlock) {
+    _.each(arrOfCanonicalBasicBlock, function(x){
+        var targetNode;
+        var lastInst = lastInstruction(asmArr, x);//asmArr[x.end-1].text;
+        switch(instructionType(lastInst)) {
+            case jmpInst:
+                {
+
+                //we have to deal only with jmp destination, jmp instruction are always taken.
+                    //edge from jump inst
+                    console.log("jmp");
+                    targetNode = extractNodeNameFromInstruction(lastInst);
+                    setEdge(edge, x.nameId, targetNode, 'blue');
+
+                    edges.push(JSON.parse(JSON.stringify(edge)));
+                    console.log(edge);
+                }
+                break;
+            case conditionalJmpInst:
+                {
+                    console.log("condit jmp");
+                //deal with : branche taken, branch not taken
+                    targetNode = extractNodeNameFromInstruction(lastInst);
+                    setEdge(edge, x.nameId, targetNode, 'green');
+                    edges.push(JSON.parse(JSON.stringify(edge)));
+                    console.log(edge);
+
+                    targetNode = hasName(asmArr,x)? asmArr[x.end].text: generateName(x.nameId, x.end);
+                    setEdge(edge, x.nameId, targetNode, "red");
+                    edges.push(JSON.parse(JSON.stringify(edge)));
+
+                    console.log(edge);
+                }
+                break;
+            case notRetInst:
+                {
+                //precondition: lastInst is not last instruction in asmArr (but it is in canonical basic block)
+                //note : asmArr[x.end] expected to be .L*:(name of a basic block)
+                //       this .L*: has to be exactly after the last instruction in the current canocial basic block
+                    var nextNodeName = asmArr[x.end].text;
+                    setEdge(edge, x.nameId, nextNodeName, 'grey');
+                    edges.push(JSON.parse(JSON.stringify(edge)));
+                    console.log("not ret inst");
+                    console.log(edge);
+                }
+                break;
+            default:
+                    console.log("expect ret instruction or it's variants(rep ret): "+ lastInst);
+                break;
+        }
+    });
+
+    console.log(edges);
+
+    return edges;
+};
+
+
+ControlFlowGraph.prototype.generateCFGStructure = function (asmArr){
```

Comment:
I'm running out of time to review, so am going to skim the rest of the file. I think you get the idea of most of my comments :)

Reply:
done